### PR TITLE
Add Decimal.is_decimal/1 and deprecate Decimal.decimal?/1

### PR DESCRIPTION
**Pull request by wojtekmach**

### CODE REVIEW

File: lib/decimal.ex [hunk 1]
```diff
@@ -261,13 +261,35 @@ defmodule Decimal do
...
+  doc_since("1.9.0")
+  # TODO: make it similar to Kernel.is_struct when we require Elixir v1.10
+  defmacro is_decimal(term) do
+    quote do
+      case unquote(term) do
+        %Decimal{} -> true
+        _ -> false
+      end
+    end
+  end
```
Comment (ericmj) [inline lib/decimal.ex]: We should check `__CALLER__.context == :guard` and use the new map guard functions.

Comment (wojtekmach) [reply lib/decimal.ex]: I believe the new guards require more recent OTP versions. We could ship this to start people on migrating, add conditional logic to make this macro more useful on recent OTP, or wait with this change until decimal v2.0 where I assume we would require more recent Elixir. I think conditional logic or waiting is probably better, wdyt?

Comment (wojtekmach) [reply lib/decimal.ex]: @ericmj ping

Comment (ericmj) [reply lib/decimal.ex]: It's a bit confusing that it is named `is_` but doesn't work in guards. We can do conditional compilation against 1.10 and use the new is_struct / is_map_key functions. WDYT?

Comment (ericmj) [reply lib/decimal.ex]: Would also need to document that it only works as a guard on elixir 1.10.

**Comment by ericmj** [comment]
Thank you! 💜

**Pull request merged**